A new digitization model for the ETL detector

parbol opened the pull request:
#### PR description:

This PR implements a new digitization model for the MTD ETL detector. This new model has been presented several times in the MTD DPG meetings. You can find a summary at [1]. 

This PR introduces potential changes in the digi efficiency (since thresholds have been slightly changed), and in the time of arrival of the charged particles (since the time resolution model has been updated). 

The Digi structure (ETLSample) has been also updated to accomodate for one additional parameter: the time over threshold. The ETLUncalibratedRecHit has also been modified maintaining backwards compatibility. If the time over threshold is zero, meaning that it's not present in the DIGI, then the old local reconstruction is run. On the contrary, if the time over threshold is not zero, a new local reconstruction is run, including the time-walk calibration.

[1] https://indico.cern.ch/event/1254512/contributions/5270127/attachments/2595708/4480662/DPG-MTD-Feb17.pdf

#### PR validation:

1.- A full ttbar workshop has been run up to the DQM stage and the output of the MTD DQM has been inspected.
2.- The same has been done starting from old digis to make sure the code is backwards compatible.
3.- The tests suggested in PRWorkflow.html have been run.

parbol committed:
A new digitization model is proposed for the ETL detector

parbol committed:
Adapted ETLsample to have a constructor with only toa but setting tot at 0

parbol committed:
Missing committed file

parbol committed:
Added missing file

parbol committed:
Fix parameters in uncalibrated rechits

parbol committed:
Update ETLSample and FTLUncalibratedHIts for bugs

parbol committed:
Passed code-checks and code-format

parbol committed:
Updated classes_def.xml for ETLSample

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"348fc014aa876cf4a03c827727bd50df72cac8e4":{"files":["SimFastTiming/FastTimingCommon/src/ETLPulseShape.cc"],"time":1705604495},"3f54e8ee02110956aa0628e8c15ab1b5f0234e66":{"files":["RecoLocalFastTime/FTLRecProducers/python/mtdUncalibratedRecHits_cfi.py"],"time":1705611299},"4b7a73dde41a855ad968594fb9e849a9b1edb364":{"files":["DataFormats/FTLDigi/src/classes_def.xml"],"time":1705858050},"6013910aadb074710a8f8935e5df1f95ffc61b35":{"files":["DataFormats/FTLDigi/interface/ETLSample.h","RecoLocalFastTime/FTLCommonAlgos/plugins/ETLUncalibRecHitAlgo.cc","SimFastTiming/FastTimingCommon/interface/ETLDeviceSim.h","SimFastTiming/FastTimingCommon/interface/ETLElectronicsSim.h","SimFastTiming/FastTimingCommon/interface/MTDShapeBase.h","SimFastTiming/FastTimingCommon/src/ETLDeviceSim.cc","SimFastTiming/FastTimingCommon/src/ETLElectronicsSim.cc","SimFastTiming/FastTimingCommon/src/ETLPulseShape.cc","SimFastTiming/FastTimingCommon/src/MTDShapeBase.cc"],"time":1705846805},"74e001df087b25905683c3b6d27b3dedffc29b32":{"files":["DataFormats/FTLDigi/interface/ETLSample.h","RecoLocalFastTime/FTLCommonAlgos/plugins/ETLUncalibRecHitAlgo.cc","SimFastTiming/FastTimingCommon/interface/ETLElectronicsSim.h","SimFastTiming/FastTimingCommon/src/ETLDeviceSim.cc","SimFastTiming/FastTimingCommon/src/ETLElectronicsSim.cc"],"time":1705846401},"923b64a9ab5e6de7a7ac900b8da82f2c085e85d3":{"files":["DataFormats/FTLDigi/interface/ETLSample.h","RecoLocalFastTime/FTLCommonAlgos/plugins/ETLUncalibRecHitAlgo.cc"],"time":1705578791},"97ff839f1b451fb07ba7d22998df7538aa41718c":{"files":["SimFastTiming/FastTimingCommon/interface/ETLPulseShape.h"],"time":1705595794},"e950e7e51cd61a65dd7a69de9f7bfdb99592fe43":{"files":["DataFormats/FTLDigi/interface/ETLSample.h","RecoLocalFastTime/FTLCommonAlgos/plugins/ETLUncalibRecHitAlgo.cc","RecoLocalFastTime/FTLRecProducers/python/mtdUncalibratedRecHits_cfi.py","SimFastTiming/FastTimingCommon/interface/ETLDeviceSim.h","SimFastTiming/FastTimingCommon/interface/ETLElectronicsSim.h","SimFastTiming/FastTimingCommon/interface/MTDShapeBase.h","SimFastTiming/FastTimingCommon/python/mtdDigitizer_cfi.py","SimFastTiming/FastTimingCommon/src/ETLDeviceSim.cc","SimFastTiming/FastTimingCommon/src/ETLElectronicsSim.cc","SimFastTiming/FastTimingCommon/src/MTDShapeBase.cc"],"time":1705576922},"emoji":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38482

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @parbol  for master.

It involves the following packages:

- DataFormats/FTLDigi (**simulation, upgrade**)
- RecoLocalFastTime/FTLCommonAlgos (**reconstruction, upgrade**)
- RecoLocalFastTime/FTLRecProducers (**reconstruction, upgrade**)
- SimFastTiming/FastTimingCommon (**simulation, upgrade**)


@srimanob, @civanch, @mdhildreth, @mandrenguyen, @subirsarkar, @jfernan2, @cmsbuild can you please review it and eventually sign? Thanks.
@fabiocos, @missirol, @rovere this is something you requested to watch as well.
@antoniovilela, @sextonkennedy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fabiocos commented on DataFormats/FTLDigi/src/classes_def.xml:
```diff
@@ -7,7 +7,9 @@
    <class name="std::vector<BTLSample>"/>
-   <class name="ETLSample"/>
+   <class name="ETLSample" ClassVersion="10">
+     <version ClassVersion="10" checksum="1290724074"/>
```
@parbol according to my understanding of the instructions linked in the twiki https://github.com/cms-sw/cmssw/pull/28349/files#r343383848 we should have a class version corresponding to the original ```ETLSample``` and then another one, with a differet checksum, for the updated version, with the additional member

parbol committed:
Corrected classes_def.xml

fabiocos commented:
please test workflow 27634.0

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38483

- This PR adds an extra 68KB to repository

cmsbuild commented:
Pull request #43762 was updated. @mdhildreth, @srimanob, @mandrenguyen, @subirsarkar, @civanch, @jfernan2 can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-878eaf/36954/summary.html
**COMMIT**: acd22a614e66e7236dd4b1d611533653c73554f1
**CMSSW**: CMSSW_14_0_X_2024-01-21-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43762/36954/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-21-2300+878eaf/60676/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 14250 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-21-2300+878eaf/60676/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3345055
 - DQMHistoTests: [Total failures: 36693](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-21-2300+878eaf/60676/validateJR.html)
 - DQMHistoTests: Total nulls: 15
 - DQMHistoTests: Total successes: 3308325
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 165 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-21-2300+878eaf/60676/triggerResults): found differences in 4 / 47 workflows

fabiocos commented:
both BTL and ETL digitization algorithms are using the ```mix``` random number engine, so changing ETL use of random numbers should affect whatever else is using that random number sequence. Nevertheless, fixing the missing initialization is for sure appropriate.

parbol committed:
The fallTime_ variable has been initialized to zero in the constructor of MTDBaseShape.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38513

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38513/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38513/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

parbol committed:
Passed code-format and code-checks again

cmsbuild commented:
Pull request #43762 was updated. @jfernan2, @mandrenguyen, @civanch, @cmsbuild, @mdhildreth, @srimanob, @subirsarkar can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-878eaf/36999/summary.html
**COMMIT**: 174ad95ca86e72f2628dac42db97a84fac8617a4
**CMSSW**: CMSSW_14_0_X_2024-01-23-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43762/36999/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 102 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-23-1100+878eaf/60705/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 11277 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-23-1100+878eaf/60705/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3247526
 - DQMHistoTests: [Total failures: 36532](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-23-1100+878eaf/60705/validateJR.html)
 - DQMHistoTests: Total nulls: 12
 - DQMHistoTests: Total successes: 3210960
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-01-23-1100+878eaf/60705/triggerResults): found differences in 3 / 46 workflows

civanch commented:
+1

regression is not expected

fabiocos commented:
@parbol there are some plots in the MTD Tracks checks that seem empty, could you please cross check also ETL DIGI and local reco plots? Are all the changes as expected?

fabiocos commented:
I looked at 2k TTbar events (wf 24834.0) to cross verify the short matrix test plots with higher statistics. There is clearly an issue breaking the cluster - track matching efficiency in ETL:
![TPEtaMatchEff](https://github.com/cms-sw/cmssw/assets/4058194/d64db9f8-bae9-4814-8484-314ad5f736a5)
This needs to be understood and fixed, before this code is punt in production.

fabiocos commented:
type mtd

fabiocos commented:
Looking at the local position plots, there seems to be an issue with the definition and propagation of the row and column info ( to be verified).

parbol commented:
I will work on this as soon as possible. The row and column definition were not changed in this update (at least not on purpose). I will perform further tests asap and let you know.

parbol commented:
Hi Fabio,

I have been working on this over the weekend and I have found the problem (although I am afraid I think this PR needs additional studies). Let me explain:

The main problem I identified was in a method in the ETLElectronicsSim class: in order to allow backwards compatibility I am using the ToT as a marker. Since the previous version of the digitization didn't have a ToT, when reading old Digis, ToT takes the value 0 and then the ETLUncalibratedRecHit applies the old local reconstruction. This works, only if in the new digitization ToT cannot be 0. This is not a problem since a ToT of zero means that the hit should not be reconstructed. The problem is that I introduced an if statement in such a way that if this was the case the ETLSample was not inserted in the ETLDataFrame and this was essentially making a mess.

I have corrected that and run the whole validation over 400 events for the new digitiztion and the old. You can actually see the DQM files at:

https://cernbox.cern.ch/s/Ci1YZPqAshCnWBt
https://cernbox.cern.ch/s/N89dHwZtwGXAQBB

Let me summarize my findings:

At digi and local reco level:

+ The energy spectrum is different (and this is actually expected since this has changed in the code)
+ The time distributions are very similar, being a bit broader for the new digitization (this can be adapted)
+ Occupancy plots and X, Z, Y distributions look similar for the new and old
+ The number of events is much lower in the new digitization. This could be making sense since the thresholds are now more strict, plus there are additional conditions (for instance regarding the signal shape) that were not before. On the other hand, I find the difference too large and I think this deserves more investigation.

At tracking level:

+ I see a clear deterioration of the ETL matching efficiency. 

Because of these reasons I believe this needs to be further investigation  (which I'll do in the incoming days).

fabiocos commented:
@parbol thanks, so I would suggest to put this PR on hold, pending further clarification, and follow-up offline about the various studies with @cms-sw/mtd-dpg-l2

parbol committed:
Several bug fixes applied on the MPV and the landau noise spread

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43762/38685

- This PR adds an extra 72KB to repository

fabiocos commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-878eaf/37199/summary.html
**COMMIT**: db6029c730b77d2c592cff2e051dea4fa4e354b1
**CMSSW**: CMSSW_14_0_X_2024-02-03-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43762/37199/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 7 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-03-1100+878eaf/60874/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 11539 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-03-1100+878eaf/60874/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248614
 - DQMHistoTests: [Total failures: 36328](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-03-1100+878eaf/60874/validateJR.html)
 - DQMHistoTests: Total nulls: 12
 - DQMHistoTests: Total successes: 3212252
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-03-1100+878eaf/60874/triggerResults): found differences in 3 / 46 workflows

fabiocos commented:
a quick cross check on 2k TTbar events (wf. 24834.0) confirms that the observed issues seem gone. @cms-sw/simulation-l2 @cms-sw/reconstruction-l2 please resume the review

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_1_X. Please open a backport if it should also go in to CMSSW_14_0_X.

jfernan2 commented:
+1

civanch commented:
+1

fabiocos commented:
@cms-sw/upgrade-l2 any objection to the integration of this code? It missed unfortunately 14_0_X, it would be useful to validate is at the first occasion

subirsarkar commented:
+Upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1

makortel commented:
It seems like this PR would be causing crashes in some multithreaded Phase2 workflows, see https://github.com/cms-sw/cmssw/issues/43942

makortel commented on SimFastTiming/FastTimingCommon/src/ETLDeviceSim.cc:
```diff
@@ -56,22 +74,54 @@ void ETLDeviceSim::getHitsResponse(const std::vector<std::tuple<int, uint32_t, f
-    if (!topo.isInPixel(simscaled)) {
-      LogDebug("ETLDeviceSim") << "Skipping hit ouf of pixel # " << hitidx << " DetId " << etlid.rawId() << " tof "
-                               << toa;
-      continue;
```
It seems like the removal of this `continue` when `topo.isInPixel(simscaled)` could be causing the out-of-bounds crash investigated in https://github.com/cms-sw/cmssw/issues/43942

fabiocos replied:
@makortel thanks for the investigation. ETL row-colum is assigned here, based on the https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDGeometryBuilder/src/RectangularMTDTopology.cc class. On the other hand the statement was modified so as to take into account a possible signal also in the interpixel gaps, i.e. the non-fully-active part of the logical partition of the surface of the detector. BTW the wrong definition of the gap size was responsible for a reduced efficiency, and was recently fixed (.i.e. too may simhits were skipped).

The point is to my mind s whether removing that ```continue``` can produce unphysical row-column values when ```topo.pixel(simscaled)``` is called. Triggered by https://github.com/cms-sw/cmssw/issues/43942#issuecomment-1941802086 , I added an assert and printed the maximum allowed range in the MTD clusterizer https://github.com/cms-sw/cmssw/blob/master/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73 , and this shows 

```
Buffer size [33,17]
```

i.e. 32,16 should be an allowed value that does not trigger the assert. Anyway since the assertion fires, there is clearly something wrong, by reproducing the crash on PU jobs we should be able to finally catch it.

makortel replied:
> I added an assert and printed the maximum allowed range in the MTD clusterizer https://github.com/cms-sw/cmssw/blob/master/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73 , and this shows
> 
> ```
> Buffer size [33,17]
> ```
> 
> i.e. 32,16 should be an allowed value that does not trigger the assert.

The printout
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/src/MTDThresholdClusterizer.cc#L73
adds 1 to both the number of rows and columns. The `MTDArrayBuffer::setSize()`, however, uses the number of rows and columns directly for the space
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L134-L143
and otherwise seems to use the usual 0-based indexing
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L170-L179
https://github.com/cms-sw/cmssw/blob/db9aceb6f549d8e8352d81ae5ec1e511af6f38bc/RecoLocalFastTime/FTLClusterizer/interface/MTDArrayBuffer.h#L108
so I'd be tempted to conclude the buffer has size for 32 x 16 elements, i.e. ranges of `[0, 31] x [0, 15]`.

fabiocos replied:
@makortel you are correct and I was wrong. My bad, I should have cross checked earlier with the geometry tests I have setup, in particular https://github.com/cms-sw/cmssw/blob/master/Geometry/MTDGeometryBuilder/test/MTDDigiGeometryAnalyzer.cc
The output for scenario D098 is quite clear:

```

 Subdetector 19 MTD Det EModule_Timingactive
 Rows     32 Columns 16 ROCS X   2 ROCS Y  1 Rows/ROC 16 Cols/ROC 16 Pitch X 0.13125 Pitch Y 0.1325 Sensor Interpad X 0.005 Sensor Interpad Y 0.005 Sensor Border X 0.005 Sensor Border Y 0.005
Pixel center location:
0 0  (-2.03437,-0.99375,0) 
0 1  (-2.03437,-0.86125,0) 
0 2  (-2.03437,-0.72875,0) 
0 3  (-2.03437,-0.59625,0) 
0 4  (-2.03437,-0.46375,0) 
0 5  (-2.03437,-0.33125,0) 
0 6  (-2.03437,-0.19875,0) 
0 7  (-2.03437,-0.06625,0) 
0 8  (-2.03437,0.06625,0) 
0 9  (-2.03437,0.19875,0) 
0 10  (-2.03437,0.33125,0) 
0 11  (-2.03437,0.46375,0) 
0 12  (-2.03437,0.59625,0) 
0 13  (-2.03437,0.72875,0) 
0 14  (-2.03437,0.86125,0) 
0 15  (-2.03437,0.99375,0) 
1 0  (-1.90312,-0.99375,0) 
1 1  (-1.90312,-0.86125,0) 
1 2  (-1.90312,-0.72875,0) 
1 3  (-1.90312,-0.59625,0) 
1 4  (-1.90312,-0.46375,0) 
1 5  (-1.90312,-0.33125,0) 
1 6  (-1.90312,-0.19875,0) 
1 7  (-1.90312,-0.06625,0) 
1 8  (-1.90312,0.06625,0) 
1 9  (-1.90312,0.19875,0) 
1 10  (-1.90312,0.33125,0) 
1 11  (-1.90312,0.46375,0) 
1 12  (-1.90312,0.59625,0) 
1 13  (-1.90312,0.72875,0) 
1 14  (-1.90312,0.86125,0) 
1 15  (-1.90312,0.99375,0) 
2 0  (-1.77187,-0.99375,0) 
2 1  (-1.77187,-0.86125,0) 
(...)
30 14  (1.90312,0.86125,0) 
30 15  (1.90312,0.99375,0) 
31 0  (2.03437,-0.99375,0) 
31 1  (2.03437,-0.86125,0) 
31 2  (2.03437,-0.72875,0) 
31 3  (2.03437,-0.59625,0) 
31 4  (2.03437,-0.46375,0) 
31 5  (2.03437,-0.33125,0) 
31 6  (2.03437,-0.19875,0) 
31 7  (2.03437,-0.06625,0) 
31 8  (2.03437,0.06625,0) 
31 9  (2.03437,0.19875,0) 
31 10  (2.03437,0.33125,0) 
31 11  (2.03437,0.46375,0) 
31 12  (2.03437,0.59625,0) 
31 13  (2.03437,0.72875,0) 
31 14  (2.03437,0.86125,0) 
31 15  (2.03437,0.99375,0) 

```

as I was initially suggesting in https://github.com/cms-sw/cmssw/issues/43942#issuecomment-1943801653 . The clusterizer buffer is larger than the allowed real range, this will have to be possibly reviewed, and in any case adding an assert check is useful to catch immediately possible issues. Even in the ```RectangularMTDTopology``` in first place. 
I'll check how to preserve the treatment of the inter-pixel active area gap that @parbol has added without spoiling the index of the pixel itself.
